### PR TITLE
adding support for promises

### LINKED ISSUE #1306: Support for promises
Feel free to rename this issue if it's something more specific. This started in this comment: https://github.com/facebook/prepack/issues/1285#issuecomment-354828615 and I've found a reduced test case which I'm filing here.

According to Sebastian's comment in that issue, evaluating promises should work, but I can't get even a basic program working. The simplest program throws the error described in the other issue, "all native function values should be intrinsics". Here's the program:

https://prepack.io/repl.html#GYVwdgxgLglg9mABAdwIYygCjASkQbwChFEAnAUyhFKTHOUQAVS4BbGAZ3Mwo7gBsAbuUQBeAHyIuUACoxW5OCCy8BwgDSJcOANyEAvoUJoMmAIwAGKzgB0UABbkwmCAj79yN-nADmmAOT2MP44OEA

### PR DESCRIPTION
**Pull request by sendilkumarn**
This tries to close #1306

### CODE REVIEW

**Comment by NTillmann** [comment]
This is going to be a bit more tricky...

We should start from some kind of test case (in test/serializer), some minimal functionality we want Prepack to support around properties. Something like observing the resolved result of `p = new Promise(resolve => resolve(42))`, e.g.
```js
(function(){
  let p = new Promise(resolve => resolve(42));
  inspect = function() { return p; }
})();
```
Unfortunately, observing the result of a promise is only possible via `then` chaining, which is not executed immediately. So we probably also need to extend the existing test-runner.js to allow us to inspect the result in the prepacked heap.

**Comment by sendilkumarn** [comment]
@NTillmann  I agree that this is tricky. 

>need to extend the existing test-runner.js 

Thinking whether we need this, the test cases seems to run without any issues

File: src/values/ObjectValue.js [hunk 1]
```diff
@@ -367,7 +367,8 @@ export default class ObjectValue extends ConcreteValue {
     if (this.$WeakSetData !== undefined) return "WeakSet";
     if (isReactElement(this) && this.$Realm.react.enabled) return "ReactElement";
     if (this.$TypedArrayName !== undefined) return this.$TypedArrayName;
-    // TODO #26 #712: Promises. All kinds of iterators. Generators.
+    if (this.$Promise !== undefined) return this.$Promise;
```
Comment (hermanventer) [inline src/values/ObjectValue.js]: Shouldn't this return "Promise" rather than this.$Promise?

**Comment by hermanventer** [comment]
Please fix the Flow errors.

**Comment by sendilkumarn** [comment]
@hermanventer sorry I missed this completely. 

@trueadm I would love to have some help with this. Is there any doc or wiki available for this?

**Comment by NTillmann** [comment]
To answer your question: Not really; is there anything in particular you'd like to get help with?

Closing due to lack of activity; feel free to reopen.

**Pull request closed without merging**